[CIVIS-8000] Add ITR how it works page (#20729)
DataDog/documentation: content/en/continuous_integration/intelligent_test_runner/how_it_works.md

---
title: How It Works
kind: documentation
further_reading:
  - link: "https://www.datadoghq.com/blog/streamline-ci-testing-with-datadog-intelligent-test-runner/"
    tag: "Blog"
    text: "Streamline CI testing with Datadog Intelligent Test Runner"
  - link: "https://www.datadoghq.com/blog/monitor-ci-pipelines/"
    tag: "Blog"
    text: "Monitor all your CI pipelines with Datadog"
---

## What is Intelligent Test Runner?

Intelligent Test Runner is Datadog's test impact analysis solution. Test impact analysis is a technique that has been getting popular in the past couple of decades, but is usually hard and time-consuming to implement. Intelligent Test Runner cuts through that complexity with a simple setup process.

> Test impact analysis maps each test to the set of code files in your repository that your test exercises, with the goal of reducing the time taken by the testing phase of your CI. This is achieved by analyzing the code changes you are trying to test, and skipping the tests that do not impact your code changes.

Reviewer: > Test impact analysis maps each test to the set of code files in your repository that your test exercises, with the goal of reducing the time taken by the testing phase of your CI.

It feels like the phrase suggests that mapping tests to sets of files is going to reduce testing time somehow. I'd  suggest elaborating a bit, something along these lines:

"... with the goal of skipping the tests that are not affected by the code changes that are being introduced, in order to reduce the time taken by the testing phase of your CI"

An extreme example of this would be a Pull Request that is only changing a typo on a README file. For that PR, running all tests doesn't provide much value. On the contrary, if you have any flaky tests your CI might fail and you might have to retry several times before being able to merge. This is a waste of both developer and CI time. With Intelligent Test Runner, a PR that is only changing a README file would skip all tests.

## What sets it apart

> Some test selection solutions try to make up for the lack of code coverage information by trying to infer in a probabilistic fashion which tests are relevant through the use of Machine Learning. This results in a test selection mechanism that can skip tests that were relevant, and cause build failures in your default branch.

Reviewer 2: is it the first time we mention "code coverage" in this section? It seems we were referring to it as "test impact analysis". It might surprising for users to read code coverage now 

Maybe this?
```suggestion
Some test selection solutions make up for the lack of per test impact analysis through the use of Machine Learning techniques, which infer what tests are relevant for a given code change in a probabilistic fashion. This results in a test selection mechanism that can skip tests that were relevant, causing build failures in your default branch.
```

Reviewer 3: @juan-fernandez I like the change. Think we should also add this other aspect at the end? 

"ML-based techniques also typically require longer periods of data collection before they're able to work. Intelligent Test Runner begins working immediately after a baseline of code coverage is gathered."

> Other test impact analysis solutions are based off code coverage but are hard to set up. Additionally, they only look at the diff in your latest commit to evaluate which tests to run. This means that you need to guarantee that all your commits run through CI. Even if you do, this doesn't play well with GitHub's Pull Requests, which only take into account the CI status of the latest commit to allow merging the PR.

Reviewer 2: I'm fine with using code coverage (actually maybe we should be more explicit), but if that's the case, we might have to "define it" earlier 

> Even if you do, this doesn't play well with GitHub's Pull Requests, which only take into account the CI status of the latest commit to allow merging the PR.

I'm not sure I understand this 🤔

Reviewer 2: not great either but maybe an idea: 
```suggestion
Other test impact analysis are based off code coverage too but only consider the last commit diff to evaluate which tests to run. As an example, this is a problem with GitHub's Pull Requests, that only take into account the CI status of the latest commit to allow merging. As a result, you either run all your commits through CI, or you have the risk of skipping a test that should've run. 
```

Reviewer 3: Look good, just a slight word change:

"While others calculate test impact analysis using code coverage too, they only consider the last commit diff when evaluating which tests to run. As an example, this is a problem with GitHub's Pull Requests, which only take into account the CI status of the latest commit to allow merging. As a result, you must run all commits through CI or risk skipping tests that should have run."

Intelligent Test Runner leverages per-test code coverage information along with data from [Test Visibility](../../tests/) to search previous test in all relevant past commits. Configuration of Intelligent Test Runner is a one-click operation on most languages, and the results are accurate and more precise than other methods.

> ## How test selection works

Author: Good feedback, thanks. I'll try to revive my attempts at making a git-centric diagram, which I think provide more information.

> ## How test selection works

Reviewer 2: is this still a concern @nikita-tkachenko-datadog ? Or the git centric diagram helps answering that?

> When you enable Intelligent Test Runner, you transparently start collecting per-test (or per-suite, depending on the language) code coverage that is sent to Datadog.

Reviewer 2: ```suggestion
When you enable Intelligent Test Runner, per-test (or per-suite, depending on the framework) code coverage is transparently collected and sent to Datadog.
```

> When starting a new test session in your CI, the Datadog instrumentation requests a list of skippable tests from the backend. It provides information about the current commit, the service that is being tested, and OS and runtime configuration. The Datadog backend uses that information to search for previous test runs that can be used as proof that, if a specific test is run, we wouldn't be gaining any value. This happens whenever a previous passed test run is found that was run on the same set of impacted and [tracked](../#tracked-files) files.

Author: > This happens whenever a previous passed test run is found that was run on a commit that has the same version of the covered and [tracked](../#tracked-files) files.

Is that better?

> When starting a new test session in your CI, the Datadog instrumentation requests a list of skippable tests from the backend. It provides information about the current commit, the service that is being tested, and OS and runtime configuration. The Datadog backend uses that information to search for previous test runs that can be used as proof that, if a specific test is run, we wouldn't be gaining any value. This happens whenever a previous passed test run is found that was run on the same set of impacted and [tracked](../#tracked-files) files.

Reviewer: Yes, this is much easier to understand

> When starting a new test session in your CI, the Datadog instrumentation requests a list of skippable tests from the backend. It provides information about the current commit, the service that is being tested, and OS and runtime configuration. The Datadog backend uses that information to search for previous test runs that can be used as proof that, if a specific test is run, we wouldn't be gaining any value. This happens whenever a previous passed test run is found that was run on the same set of impacted and [tracked](../#tracked-files) files.

Reviewer 2: (just an idea) 
```suggestion
When starting a new test session in your CI, the Datadog library requests a list of skippable tests (or suites) from the backend. This request includes information about the current commit, the service that is being tested, and OS and runtime configuration. The Datadog backend uses that information to search for previous test runs that can be used as proof that, if a specific test is run, it would use exactly the same code. This happens whenever a previous passed test run is found that was run on the same set of impacted and [tracked](../#tracked-files) files.
```

{{< img src="continuous_integration/itr_test_selection.png" alt="Intelligent Test Runner Test Selection Diagram" style="width:80%;">}}

> The Datadog instrumentation then removes any skippable tests from the list that are marked unskippable in source. It then proceeds to run the tests, but directs the test framework to skip those that remain in the skippable test list.

Reviewer 2: ```suggestion
The Datadog library then removes tests marked as unskippable in source from the skippable tests list. It then proceeds to run the tests, but directs the test framework to skip those that remain in the skippable test list.
```

---
title: How It Works
kind: documentation
further_reading:
  - link: "https://www.datadoghq.com/blog/streamline-ci-testing-with-datadog-intelligent-test-runner/"
    tag: "Blog"
    text: "Streamline CI testing with Datadog Intelligent Test Runner"
  - link: "https://www.datadoghq.com/blog/monitor-ci-pipelines/"
    tag: "Blog"
    text: "Monitor all your CI pipelines with Datadog"
---

> ## What is Intelligent Test Runner?

Reviewer 3: re: warning - it is using sentence-style capitalization. Intelligent Test Runner is the proper name of the product.

Intelligent Test Runner is Datadog's test impact analysis solution. Test impact analysis is a technique that has been gaining popularity over the past few decades. However, it's typically hard and time-consuming to implement. Intelligent Test Runner simplifies this complexity with an easy setup process.

> Test impact analysis maps each test to the set of code files in your repository that your test uses. Its goal is to skip tests that are not affected by the code changes that are being introduced, in order to reduce the time taken by the testing phase of your CI.

Reviewer 3: re: warning
change last sentence to, 
"Its goal is to skip tests that are not affected by the code changes that are being introduced. This leads to a direct reduction in time spent testing in CI.

An extreme example is a Pull Request that only changes a typo on a README file. For that PR, running all tests doesn't provide much value. On the contrary, flaky tests might make your CI fail, forcing you to retry the pipeline, potentially multiple times, before merging. This is a waste of both developer and CI time. With Intelligent Test Runner, a PR changing a README file would skip all tests.

## What sets it apart

Some test selection solutions try to make up for the lack of code coverage information by trying to infer in a probabilistic fashion which tests are relevant through the use of Machine Learning. This results in a test selection mechanism that can skip tests that were relevant, and cause build failures in your default branch.

> Other test impact analysis solutions are based off code coverage but are hard to set up. Additionally, they only look at the diff in your latest commit to evaluate which tests to run. This means that you need to guarantee that all your commits run through CI. Even if you do, this doesn't play well with GitHub's Pull Requests, which only take into account the CI status of the latest commit to allow merging the PR.

Reviewer 3: re: warning - change last sentence:

"Even if they do, there still may be issues related to GitHub's Pull Requests. When it comes to merging the PR, only the CI status of the latest commit is taken into account."

Intelligent Test Runner leverages per-test code coverage information along with data from [Test Visibility](../../tests/) to search previous test in all relevant past commits. Configuration of Intelligent Test Runner is a one-click operation on most languages, and the results are accurate and more precise than other methods.

## How test selection works

When you enable Intelligent Test Runner, per-test (or per-suite, depending on the framework) code coverage is transparently collected and sent to Datadog.

> When starting a new test session in your CI, the Datadog library requests a list of skippable tests from the backend. This request includes information about the current commit, the service that is being tested, and OS and runtime configuration. The Datadog backend uses that information to search for previous test runs that can be used as proof that, if a specific test is run, we wouldn't be gaining any value. This happens whenever a previous passed test run is found that was run on a commit that has the same version of the covered and [tracked](../#tracked-files) files.

Reviewer 3: Change wording:

"The Datadog backend uses that information to search through previous test runs to determine if a given test can be skipped. If we have a record of the test passing in a commit where the covered and tracked files are identical to the current commit, we can skip the test. We know that test wasn't impacted by the code change."

{{< img src="continuous_integration/itr_test_selection.png" alt="Intelligent Test Runner Test Selection Diagram" style="width:60%;">}}

The Datadog library then removes tests marked as unskippable in source from the skippable tests list. It then proceeds to run the tests, but directs the test framework to skip those that remain in the skippable test list.

{{< img src="continuous_integration/itr_skipped_test_run.png" alt="Intelligent Test Runner Skipped Test" style="width:80%;">}}

Let's take a look at a specific example:

{{< img src="continuous_integration/itr_example.png" alt="Intelligent Test Runner Example Diagram" style="width:80%;">}}

In the previous diagram shows a developer branch that branches out from `main` and has several commits. On each commit, the CI has been running two tests (A and B) with different results.

- **Commit 1** ran both tests. This commit contained changes that affected the tracked files, and the covered files of both A and B.
- **Commit 2** ran both tests again:

>   - Test A has to be run because, although this commit did not affect test A (no changes in tracked files or covered files), there are no previous test runs that passed for Test A. Since Intelligent Test Runner cannot guarantee that the test will pass if were run, it doesn't skip it. This time, the test passes, which indicates this is a flaky test.

Reviewer: ```suggestion
  - Test A has to be run because, although this commit did not affect test A (no changes in tracked files or covered files), there are no previous test runs that passed for Test A. Since Intelligent Test Runner cannot guarantee that the test will pass if it were run, it doesn't skip it. This time, the test passes, which indicates this is a flaky test.
```

Reviewer 3: Addressed this sentence along with the use of "will"

>   - Test A has to be run because, although this commit did not affect test A (no changes in tracked files or covered files), there are no previous test runs that passed for Test A. Since Intelligent Test Runner cannot guarantee that the test will pass if were run, it doesn't skip it. This time, the test passes, which indicates this is a flaky test.

Reviewer 3: re: warning - replacing "will"

"Since Intelligent Test Runner cannot guarantee a passing status if it were run, it doesn't skip it."
